Better jar extraction

Akirathan opened the pull request:
### Pull Request Description

Simplify JAR extraction in sbt build. Fix JAR extraction caching - don't re-extract after sbt restarts.

Motivation:
- Caching breaks once sbt is restarted - all the JAR extraction is redone even if it is not necessary.
- [JARUtils.extractFilesFromJar](https://github.com/enso-org/enso/blob/b1e5da159ef0c289a550d24dbaffcaab47f70d76/project/JARUtils.scala#L12-L199) has become a complicated, unmaintainable beast.

### Important Notes

- Every JAR to be extracted will be wrapped in a new separate project, like [opencv-thin](https://github.com/enso-org/enso/blob/b3725754c0830512ebb6c64ec7e36431fc78ccb3/build.sbt#L4884-L4893)
- Such wrapper project:
  - Enables [JarExtractPlugin](https://github.com/enso-org/enso/blob/939c052e6c26c3cc780c0af4caa4a36e8ad442f0/project/JarExtractPlugin.scala)
  - Has single input jar
  - Has two outputs - a thin jar without extracted files and a directory with extracted files.
  - Has a definition of `jarExtractor` that describes how the jar should be extracted.
  - Input jar is never modified

- To get path to the directory with extracted files: [opencv-thin/extractedFilesDir](https://github.com/enso-org/enso/blob/b3725754c0830512ebb6c64ec7e36431fc78ccb3/build.sbt#L4926-L4927)
- To get path to the thin jar: [opencv-thin/thinJarOutput](https://github.com/enso-org/enso/blob/b3725754c0830512ebb6c64ec7e36431fc78ccb3/build.sbt#L4928)
- Both tasks `extractedFilesDir` and `thinJarOutput` do extraction only if one of these conditions is true:
  - extracted files directory does not exist
  - thin jar output does not exist
  - input jar has been modified
  - Implemented in [JarExtractPlugin.extract](https://github.com/enso-org/enso/blob/939c052e6c26c3cc780c0af4caa4a36e8ad442f0/project/JarExtractPlugin.scala#L97-L121)

#### Example of extraction declaration
https://github.com/enso-org/enso/blob/5054951f3e8f4d254149224b97c3888e9afaf283/build.sbt#L4899-L4915
 
### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
Remove duplicated condition

Commit:
Move osName and arch into Platform source

Commit:
Implement JarExtract plugin

Commit:
std-image depends on opencv-thin project

Commit:
JarExtractPlugin uses directory for extracted files.

Does not enumerate them.
- Also cleans the directories properly

Commit:
Put opencv-thin in enso aggregate

Commit:
StdBits.copyDependencies simply copies native lib dirs if necessary

Akirathan commented:
The `JarExtraPlugin` also enable us to greatly simplify definition of projects that are doing jar extraction. Just look at the definition of `std-image` project in this PR and compare it to develop. If all the projects will be migrated to this new plugin, we can remove the [AnalysisOfExtractedNativeLibs](https://github.com/enso-org/enso/blob/62797fed7cb245abbc83c45fd6fe292687387bac/project/AnalysisOfExtractedNativeLibs.scala) magic.

Akirathan commented:
@JaroslavTulach @hubertp Please look at this PR and express your opinions. If you are OK with this, I will rewrite the rest of the projects to use this new `JarExtractPlugin`.

Commit:
Decrease verbosity of logging

JaroslavTulach commented on std-bits/opencv-thin/README.md:
I'd be expecting this wrapper project next to other wrappers like 
- lib/java/akka-wrapper  
- lib/java/fansi-wrapper  
- lib/java/jna-wrapper  
- lib/java/language-server-deps-wrapper  
- lib/java/poi-wrapper  
- lib/java/scala-libs-wrapper  
- lib/java/zio-wrapper

Akirathan replied:
I don't have a strong opinion on that. So I have moved `opencv-thin` next to other wrapper projects in 5054951f3e8f4d254149224b97c3888e9afaf283

JaroslavTulach commented on project/JARUtils.scala:
```diff
@@ -93,7 +93,7 @@ object JARUtils {
                       dynamicLibs = destFile :: dynamicLibs
                       if (
-                        destFile.exists() && destFile.exists() && inputJarFile
+                        destFile.exists() && inputJarFile
```
Yes, one check is enough ;-)

JaroslavTulach commented on project/JarExtract.scala:
```diff
@@ -0,0 +1,157 @@
+    )
+  }
+
+  val openCVExtractor: JarVisitor = (entry: JarEntry) => {
```
- it'd be better if this logic was part of  `build.sbt`
   - people don't want to read code of plugins
   - just configure `<plugin><configure>...</configure></plugin>` in Maven terminology 
- but in some more simplified, declarative form
- maybe some combination of regexps...

Akirathan replied:
I believe that the declaration implemented in f9b497e3d1dea22b2f0c7ced6c90d268c4a6052c is better.

JaroslavTulach commented on build.sbt:
```diff
@@ -4879,6 +4881,17 @@ lazy val cleanPolyglotRoot = taskKey[Unit](
+    libraryDependencies ++= Seq(
+      "org.openpnp" % "opencv" % opencvVersion
+    ),
+    inputJar := "org.openpnp" % "opencv" % opencvVersion,
```
- Why not read `libraryDependencies` directly instead of `inputJar`?
- maybe fail when there is more than one dependency (we don't need more than one right now)

Akirathan replied:
We need `java.io.File`. We need `Path` to the local file system. `libraryDependencies` does not give that to us. Unfortunately, there is no way to access library resolution of sbt, so you have to do that manually.

JaroslavTulach commented on build.sbt:
```diff
@@ -4879,6 +4881,17 @@ lazy val cleanPolyglotRoot = taskKey[Unit](
+      "org.openpnp" % "opencv" % opencvVersion
+    ),
+    inputJar := "org.openpnp" % "opencv" % opencvVersion,
+    jarExtractor := JarExtract.openCVExtractor
```
- good for now, but I'd like to see the (declarative) configuration here
- why not:
```
polyglotLib["amd64"] := Seq(
  ".*/windows/x86_64/.*.dll", 
  ".*/linux/x86_64/.*.so", 
  ".*/osx/x86_64/.*.dylib"
)
polyglotLib["aarch64"] := Seq(".*/osx/aarch64/.*.dylib")
```
- either as regex or glob patter
- those `polyglotLib` resources would automatically be removed from the "thin JAR"
- we could have `skipResources := Seq(....)` to drop some resources
- the rest would get into the thin JAR

Akirathan replied:
I believe that the following is better: https://github.com/enso-org/enso/blob/f9b497e3d1dea22b2f0c7ced6c90d268c4a6052c/build.sbt#L4899-L4915

JaroslavTulach reviewed (commented):
- if this improves caching behavior, then it is a win on its own, go on!
- can keep declaration of what to extract in `build.sbt`?
- e.g. replace `openCVExtractor` with a few regex or glob patterns?

Commit:
Jar extraction is specified via glob pattern mapping

Commit:
Move opencv-thin next to other wrapper projects

JaroslavTulach commented on project/JarExtractor.scala:
```diff
@@ -0,0 +1,178 @@
+
+  /** @param arch If specified, will be copied only iff the architecture is
+    *        the same as the current platform. If not specified, will be copied
+    *        to the root of the extracted files directory.
```
If specified, where the file is copied to? It is not clear from the description. Some examples would make it easier to understand.

Akirathan replied:
Example introduced in 91e13c88c76d1b3330c10025adbee4e9119e4b81 - inside https://github.com/enso-org/enso/blob/91e13c88c76d1b3330c10025adbee4e9119e4b81/project/JarExtractor.scala#L52-L53

JaroslavTulach commented on build.sbt:
```diff
@@ -4879,6 +4888,33 @@ lazy val cleanPolyglotRoot = taskKey[Unit](
+      "org.openpnp" % "opencv" % opencvVersion
+    ),
+    inputJar := "org.openpnp" % "opencv" % opencvVersion,
+    jarExtractor := JarExtractor(
```
This mapping is perfect!

JaroslavTulach commented on build.sbt:
```diff
@@ -4879,6 +4888,33 @@ lazy val cleanPolyglotRoot = taskKey[Unit](
+      "nu/pattern/opencv/linux/x86_64/*.so" -> CopyToExtractDir(
+        Some(LinuxX86_64)
+      ),
+      "nu/pattern/opencv/osx/ARMv8/*.dylib" -> CopyToExtractDir(
```
- I don't like the name `CopyToExtractDir`
- `Some(...)` is (needlessly) repeated again and again
- why not just:

```suggestion
      "nu/pattern/opencv/osx/ARMv8/*.dylib" -> PolyglotLib(MacOSArm64)
```

Akirathan replied:
Good idea. Done in 91e13c88c76d1b3330c10025adbee4e9119e4b81

JaroslavTulach commented on build.sbt:
```diff
@@ -4879,6 +4888,33 @@ lazy val cleanPolyglotRoot = taskKey[Unit](
+      "nu/pattern/opencv/windows/x86_64/*.dll" -> CopyToExtractDir(
+        Some(WindowsX86_64)
+      ),
+      "nu/pattern/opencv/*.class" -> CopyToOutputJar,
```
- shouldn't `CopyToOutputJar` be the default?

Akirathan replied:
The default behavior is to skip the entry, as documented in https://github.com/enso-org/enso/blob/91e13c88c76d1b3330c10025adbee4e9119e4b81/project/JarExtractor.scala#L13

JaroslavTulach commented on lib/java/opencv-thin/README.md:
- I would call it `lib/java/opencv-wrapper` just like the other library wrappers that are next to it

Akirathan replied:
Rename to `opencv-wrapper` in 7150e60d81731672bfa5e71ca8847e2a19639595

JaroslavTulach reviewed (approved):
The declarative `mapping` is way nicer than the original extractor function

Akirathan commented on build.sbt:
```diff
@@ -4913,44 +4945,13 @@ lazy val `std-image` = project
-          previousRun        = prev
+          polyglotLibDir     = Some(`image-native-libs`),
+          extractedNativeLibsDir =
+            Some((`opencv-thin` / extractedFilesDir).value),
```
This should be sequence - to have multiple wrapper thin jars in a single project

Commit:
Rename opencv-thin to opencv-wrapper

Commit:
Rename CopyToExtractDir to PolyglotLib

Commit:
opencv thin jar is not specified via providedJarNames parameter

Commit:
Fix glob pattern for some classes from opencv

Commit:
Remove unused extractNativeLibsFromOpenCV

Commit:
StdBits.copyDependencies can copy from more extractedNativeLibsDirs

Commit:
std microsoft is extracted via new plugin

Commit:
fmt

Commit:
Fix NoSuchElementException

Commit:
Decrease logging verbosity

Commit:
Runner_Spec succeeds on GraalVM EE

Commit:
Make sure std-microsoft.jar is copied into the polyglot dir with other dependencies

Commit:
Rename previous artifacts for clarity

Commit:
Move extract wrapper project next to extractNativeLibs task

Commit:
Keep module-info in jna-wrapper-extracted

Commit:
Refactor std-tableau to new extraction

Commit:
Add some wrapper projects to the enso aggregate

Commit:
fmt

Akirathan commented:
When migrating `std-google`, I encountered a bug in our current implementation: `io.grpc:grpc-netty-shaded:1.69.0` contains `1libio_*_osx_aarch_64.jnilib`, but I don't see this library extracted into `distribution/lib/*/Google/**/polyglot/lib` directory on my system. What is even more astonishing is that the whole `distribution/lib/*/Google/**/polyglot/lib` directory is empty on my system. I am wondering how is it even possible that Standard.Google works on Mac aarch64...

Akirathan commented:
Another issue is in `Standard.Microsoft`:
```
$ ls distribution/lib/Standard/Microsoft/**/polyglot/lib
libjnidispatch.jnilib  libnetty_tcnative_osx_aarch_64.jnilib
```
Note that the native libs have `.jnilib` suffix which is problematic. Because `System.mapLibrary` appends `.dylib` suffix. Note when extracting grpc, we explicitly replace `.jnilib` with `.dylib`.

Commit:
Migrate rest of the std libs to new extraction

JaroslavTulach commented on build.sbt:
```diff
@@ -13,14 +13,24 @@ import src.main.scala.licenses.{
+  CopyToOutputJar,
+  LinuxX86_64,
+  MacOSArm64,
+  MacOSX86_64,
```
`MacIntel64` - maybe. Or anything less cryptic than OSX86_64 ... ááá, the `X` three isn't OSX ... let it be ... just `MacOSArm64` looks way more cute.

Akirathan replied:
Renamed `x86_64` to `AMD64` in 8470150a37d1cef9fee9b8b2791c9f8ce9346290

JaroslavTulach commented on build.sbt:
```diff
@@ -4871,13 +4886,271 @@ lazy val `std-table` = project
+    },
+    jarExtractor := JarExtractor(
+      "META-INF/native/libnetty_tcnative_osx_aarch_64.jnilib" -> PolyglotLib(
+        MacOSArm64
```
Maybe:
```suggestion
        MacOSArm64, newExt=`dylib`
```

Akirathan replied:
I have chosen to do this silently during extraction in https://github.com/enso-org/enso/blob/d1b28a7146d35b10eeb54aa3cae92b7b787a2e73/project/JarExtractor.scala#L105-L113. Note that `.jnilib` is considered "deprecated".

Commit:
Merge branch 'develop' into wip/akirathan/better-jar-extract

# Conflicts:
#	build.sbt

Commit:
jlineNative is an excluded mod

Commit:
Fix after merge

Commit:
Add dummy README files to empty projects.

So their directories are in VCS.

Commit:
No conscrypt-openjdk-uber in google polyglot/java dir

Commit:
Rename x86_64 to AMD64.

Commit:
No conscrypt-openjdk-uber in Snowflake polyglot/java dir

Commit:
Fix UnknownIdentifierException that just throws IllegalArgumentException.

The cause parameter must extend `AbstractTruffleException`.

Commit:
Do not include META-INF/native-image/** jar entries from sqlitejdbc

There is native-image.properties that enables SqliteJdbcFeature, and we want to disable it.

Commit:
sqlite-wrapper is in enso aggregate

Commit:
jna-wrapper-extracted has jna-wrapper as input jar.

Commit:
fmt

Akirathan commented on engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/HostClassLoader.java:
```diff
@@ -201,7 +202,8 @@ Object readMember(String member) throws UnknownIdentifierException {
       var ctx = EpbContext.get(null);
       return ctx.getEnv().asHostSymbol(clazz);
     } catch (ClassNotFoundException ex) {
-      throw UnknownIdentifierException.create(member, ex);
```
As the [javadoc](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnknownIdentifierException.html#create(java.lang.String,java.lang.Throwable)) states, the throwable must extends `AbstractTruffleException`, otherwise the `create` method just fails with `IllegalArgumentException`. And that is what was happening. I have removed the throwable parameter and added a log statement instead.

JaroslavTulach replied:
OK.

GregoryTravis reviewed (approved):
For library changes

Akirathan commented:
I have done a simple QA on Linux IDE: 
- Listing Azure blob storage
- Reading SQLite table
- Writing table into hyperd format
All of that works. Let's merge and keep an eye for potential regressions.
